### DISCUSSION TITLE
Writing diagnostics into a sink (ie a Vec)

### DISCUSSION THREAD

**Discussion by cormacrelf** [zkat/miette#54] [Show and tell]
I wrote a little proof of concept to show writing Diagnostics into a sink for later use instead of a TTY.

https://github.com/cormacrelf/miette-sink/blob/master/examples/interactive.rs

It is an attempt to deal with two problems at the same time:

1. That a library crate sometimes needs to return a list of errors instead of printing them to a terminal. ~~IMO, it doesn't do this very well, because there's too much `Box<dyn Diagnostic + 'static>` for that to be a great fit. I think I'll redo this with a concrete generic type required,~~ now that we've merged `transparent` and also `forward` and it's easier to make enums that wrap other Diagnostic-implementing types. (Edit: done)
2. Separating warnings from errors in your code. That's a common thing with parsing and validating input -- sometimes you need to output a warning without `return Err(e)`, i.e. continue on after printing the warning and return an Ok. The method is very similar to `IntoDiagnostic` (it is a Result extension trait that looks like `.report(sink)?`), but the error type it produces is a `Reported<D>` with `D: Reportable`. That way you can control which errors and which diagnostic types are allowed to be reported in this fashion. If you put all your warnings in one enum and all your errors in another, then you get type-safe control over which diagnostics get "thrown" and which allow you to continue on.

It's not super polished at all, but there are some alright ideas here I think.

**Comment by zkat** [comment]
This is really cool! I definitely wanted something like this as part of `miette::Report` but you're convincing me that maybe this is something that should live in an external library instead? Or at least a separate type in the larger miette crate?

**Reply by cormacrelf** [reply]
Well it clearly can live outside, but surely that is true of a number of things in miette already.

The `Reported<D>` / `D: Reportable` combo is fun, but quite different from `Report`. I see eyre / anyhow / Report as being fundamentally about errors only, because of the weirdness of `?` for valid-but-inadvisable stuff. Last I checked it's quite difficult to write a warning to the terminal without creating a `Report`. Is there any API for this? Also if miette had both `Reported<D>` and Report, the similar terminology will be a bit confusing.

Other than that the rest can be mixed in fairly easily. VecSink is not technically necessary but the Debug impl is non-intuitive. It's worth adding so that people don't have to read the source code to the report handlers and understand all those internals to make it work; it is anyone's guess looking at miette docs/source ATM where the line of code is that prints something to the terminal.

More random thoughts: `D: Reportable` does require everybody to implement a new trait on all their things. Adding a derive macro arg for it would help -- that should come with adding default `severity = "warning"`, because currently having to ensure your severities match the containing enum is awkward and error-prone. The need for this is a bit like `#[serde(rename_all = "kebab-case")]`. Then you could make everybody's current code work without change -- the default severity is already error, so the default should probably be that a diagnostic is Reportable.

**Comment by zkat** [comment]
fyi, miette@3.0 has a `.related()` protocol method that lets you define a field that will be used as a "dumping ground" for  "related" diagnostics, which will be printed after the main one. Let me know what you think!
